### PR TITLE
8329997: Add MemorySegment::reinterpretate overloads for alignment constraints

### PR DESCRIPTION
**Pull request by minborg**
This PR proposes to add two overloads `MemorySegment::reinterpretate` to allow aligned reinterpretation.

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Integration blocker
&nbsp;⚠️ Title mismatch between PR and JBS for issue [JDK-8329997](https://bugs.openjdk.org/browse/JDK-8329997)

### Issue
 * [JDK-8329997](https://bugs.openjdk.org/browse/JDK-8329997): Add MemorySegment::reinterpret overloads for alignment constraints (**Enhancement** - P3) ⚠️ Title mismatch between PR and JBS.

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18717/head:pull/18717` \
`$ git checkout pull/18717`

Update a local copy of the PR: \
`$ git checkout pull/18717` \
`$ git pull https://git.openjdk.org/jdk.git pull/18717/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18717`

View PR using the GUI difftool: \
`$ git pr show -t 18717`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18717.diff">https://git.openjdk.org/jdk/pull/18717.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18717#issuecomment-2047465652)

### CODE REVIEW

File: src/java.base/share/classes/jdk/internal/foreign/SegmentFactories.java [hunk 1]
```diff
@@ -57,14 +58,18 @@ public class SegmentFactories {
...
+                                                           MemorySessionImpl sessionImpl,
+                                                           Runnable action) {
         ensureInitialized();
         if (action == null) {
             sessionImpl.checkValidState();
         } else {
             sessionImpl.addCloseAction(action);
         }
-        return new NativeMemorySegmentImpl(min, byteSize, false, sessionImpl);
+        return new NativeMemorySegmentImpl(Utils.alignUp(min, byteAlignment), byteSize, false, sessionImpl);
```
Comment (JornVernee) [inline src/java.base/share/classes/jdk/internal/foreign/SegmentFactories.java:72]: This is not correct. The base address of the segment should not change after a call to reinterpret. The alignment constraint should just be used for checking that the base address is somewhat sane. Instead, I think `reinterpretInternal` should do the check of the alignment of `addres()` with the given alignment constraint.

File: test/jdk/java/foreign/TestSegments.java [hunk 2]
```diff
@@ -380,12 +379,27 @@ void testReinterpret() {
...
+            assertEquals(MemorySegment.ofAddress(42).reinterpret(JAVA_INT, arena, _ -> counter.incrementAndGet()).scope(), arena.scope());
+        }
+        assertEquals(counter.get(), 3);
+    }
+
+    @Test
+    void testReinterpretAlignment() {
+        AtomicInteger counter = new AtomicInteger();
+        try (Arena arena = Arena.ofConfined()){
+            var segment = MemorySegment.ofAddress(41).reinterpret(JAVA_INT);
```
Comment (JornVernee) [inline test/jdk/java/foreign/TestSegments.java:396]: i.e. this call (and the one below) should throw an exception.

Comment (mcimadamore) [reply test/jdk/java/foreign/TestSegments.java:396]: Note - this means that the javadoc of the new methods should also document an IllegalArgumentException.

**Comment by mcimadamore** [comment]
Should we add the full set of overloads? E.g.
reinterpret(arena, cleanup)
reinterpret(size)
reinterpret(size, arena, cleanup)
reinterpret(size, align) // this is missing
reinterpret(size, align, arena, cleanup)  // this is missing
reinterpret(layout)
reinterpret(layout, arena, cleanup)

In other places (e.g. `asSlice`) we accept size/align where we also accept a layout

**Pull request closed without merging**